lib/core/src/Cardano/Pool/Performance.hs

```diff
@@ -0,0 +1,198 @@
+-- deviates slightly from it because of different real constraints:
+--
+--     a) The wallet software doesn't have a reliable access to various epoch
+--     details such as the stake distribution for past epochs.
+--
+--     b) The wallet software makes therefore a significant use of the current
+--     epoch's data which represents unfinished snapshots of epochs which, because
+--     of their non-deterministic nature, may favor leaders elected early in the
+--     process more than those not yet elected.
+--
+-- To cope with these two issues, we try to harmonize past known epochs with the
+-- ongoing one by averaging over a whole range of epochs, simultaneously. This
+-- means computing a ratio of the total blocks produced by a leader on the total
+-- number of blocks we could reasonably expect a leader to have produced, across
+-- many epochs at once (i.e., doing a ratio of sums, instead of a sum of
+-- ratios).
+--
+-- This should be less punitive for leaders that have not yet been elected as
+-- part of the ongoing epoch while having still performed reasonably okay in the
+-- past.
+module Cardano.Pool.Performance
+    ( EpochStats (..)
+    , apparentPerformance
+    , readPoolsPerformances
+
+    -- * Helpers
+    , count
+    ) where
+
+
+import Prelude
+
+import Cardano.Pool.DB
+    ( DBLayer (..) )
+import Cardano.Wallet.Primitive.Types
+    ( ActiveSlotCoefficient (..)
+    , EpochLength (..)
+    , EpochNo (..)
+    , PoolId
+    , SlotId (..)
+    , SlotNo (..)
+    )
+import Control.Monad
+    ( forM )
+import Data.Generics.Internal.VL.Lens
+    ( (^.) )
+import Data.Map.Merge.Strict
+    ( dropMissing, traverseMissing, zipWithMatched )
+import Data.Map.Strict
+    ( Map )
+import Data.Quantity
+    ( Quantity (..) )
+import Data.Word
+    ( Word64 )
+import GHC.Generics
+    ( Generic )
+import Numeric.Natural
+    ( Natural )
+
+import qualified Data.Map.Merge.Strict as Map
+import qualified Data.Map.Strict as Map
+
+data EpochStats = EpochStats
+    { poolProduction :: !Natural
+    , poolStake :: !Natural
+    , totalStake :: !Natural
+    , epochHeight :: !Natural
+    } deriving (Generic, Show)
+
+-- We consider the following variables, indexed by a given epoch
+--
+-- - n(e) = number of blocks produced in an epoch e
+-- - N(e) = number of slots in e
+-- - s(e) = stake owned by the pool in e
+-- - S(e) = total stake delegated to pools in e
+-- - ε(e) = active slot coeff, i.e. % of slots for which a leader can be elected
+--
+-- From this, we define `μ` as the _reasonably expectable block production_:
+--
+--     μ(e) =
+--       ε(e) * N(e) * s(e) / S(e)
+--
+-- Intuitively, @μ@ corresponds to the number of blocks a particular pool should
+-- have produced provided the election was exactly proportional to its relative
+-- stake.
+--
+-- This gives us the apparent performance @p@ across many epochs as:
+--
+--          i
+--          Σ n(i)
+--     p =  -----
+--          i
+--          Σ μ(i)
```

Comment:
With the absence of clamping, the performance of small pools look utterly good. For instance, on the ITN, considering only the current epoch (which is almost over, so plenty of data there), the first pool has a performance of 3265.62. 

If we look closer, we have the following stats:

```
poolProduction = 1
poolStake = 748901013
totalStake = 10102644266508743
epochHeight = 41309
```

with an active slot coeff of 0.1, we have 

```
μ = 0.1 * 41309 * 748901013 / 10102644266508743 = 0.0003062203432083028
p = 1 / μ = 3265.622360431363
```

Because this pool has a very low chance of being elected in the first place, when it does produce a block it look over performing. Although in practice, it really simply was lucky. Even if we imagine a second we had data for 14 epochs in the past, and the pool still had produced only one blocks (because the chances it would have been selected are still 4 out of 1000). It'll still have a performance of 200+ ... just because it got lucky. 

Looking at another pool that has also produced a single block but with a much greater stake like `SCAR2`: 

```
poolProduction = 1
poolStake = 891491931631
totalStake = 10102644266508743
epochHeight = 41309
```

We get a performance of ~2.74 mostly because the stake is 1200x higher. 

Perhaps, one way to solve this nicely is to take `max 1 μ`. This would clamp the first pool performance down to 1 and for scars, well, it'll clamp it to 1 in this case, but if we had 3 or more epochs of stats, it'd have no effect and give a good idea of performance. In the case of `SCAR2` for example, after 3 epochs, we would expect the pool to have produced 1.11 blocks. If it had produced only one block, then its performance would be decreased slightly.  

Also, we could make the non-production a bit less punitive by taking the square root of the sum of `μ`.

Reply:
Done. Looks much better.